src/Build/BackEnd/Components/RequestBuilder/TargetBuilder.cs

```diff
@@ -766,5 +767,45 @@ private async Task<bool> CompleteOutstandingActiveRequests(string targetName)
 
             return false;
         }
+
+        private void ComputeAfterTargetFailures(string[] targetNames)
+        {
+            // Cache intermediate results so we don't visit a target more than once.
+            var resultsWithAfterTargets = new Dictionary<string, bool>();
+
+            foreach (string targetName in targetNames)
+            {
+                if (_buildResult.ResultsByTarget.ContainsKey(targetName))
+                {
+                    // Queue of targets waiting to be processed, seeded with the specific target for which we're computing AfterTargetsHaveFailed.
+                    var targetsToCheckForAfterTargets = new Queue<string>();
+                    targetsToCheckForAfterTargets.Enqueue(targetName);
+
+                    while (targetsToCheckForAfterTargets.Count > 0)
+                    {
+                        string targetToCheck = targetsToCheckForAfterTargets.Dequeue();
+                        IList<TargetSpecification> targetsWhichRunAfter = _requestEntry.RequestConfiguration.Project.GetTargetsWhichRunAfter(targetToCheck);
+
+                        foreach (TargetSpecification afterTarget in targetsWhichRunAfter)
+                        {
+                            if (!resultsWithAfterTargets.TryGetValue(afterTarget.TargetName, out bool afterTargetFailed))
+                            {
+                                _buildResult.ResultsByTarget.TryGetValue(afterTarget.TargetName, out TargetResult result);
+                                afterTargetFailed = result != null && result.ResultCode == TargetResultCode.Failure && !result.TargetFailureDoesntCauseBuildFailure;
+
+                                resultsWithAfterTargets[afterTarget.TargetName] = afterTargetFailed;
```

Comment:
Move this into if statement?

Reply:
Originally I wanted to cache negative results as well. But you are bringing up a really good point, the logic as implemented was not correct. To really cache all results, the search would have to be depth-first so that we really have a valid result for each intermediate target. It's probably not worth adding more complexity given the expected number of targets. I have removed the cache. Thank you!